types/validator_set.go

```diff
@@ -366,6 +335,204 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// When two elements with same address are seen, the one from updates is selected.
+// Expects updates to be a list of updates sorted by address with no duplicates,
+// and to have been validated with verifyUpdatesAndComputeNewPriorities().
+// The validator's priorities in 'updates' should be already computed.
+func (vals *ValidatorSet) applyUpdates(updates []*Validator) {
+
+	existing := make([]*Validator, len(vals.Validators))
+	copy(existing, vals.Validators)
+
+	merged := make([]*Validator, len(existing)+len(updates))
+	i := 0
+
+	for len(existing) > 0 && len(updates) > 0 {
+		if bytes.Compare(existing[0].Address, updates[0].Address) < 0 {
+			merged[i] = existing[0]
+			existing = existing[1:]
+		} else {
+			merged[i] = updates[0]
+			if bytes.Equal(existing[0].Address, updates[0].Address) {
+				// validator present in both, so advance existing also
+				existing = existing[1:]
+			}
+			updates = updates[1:]
+		}
+		i++
+	}
+	for j := 0; j < len(existing); j++ {
+		merged[i] = existing[j]
+		i++
+	}
+	for j := 0; j < len(updates); j++ {
+		merged[i] = updates[j]
+		i++
+	}
+
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
+}
+
+// Checks that the validators to be removed are part of the validator set.
+// No changes are made to the validator set 'vals'.
+func verifyRemovals(deletes []*Validator, vals *ValidatorSet) error {
+
+	for _, valUpdate := range deletes {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil {
+			return fmt.Errorf("failed to find validator %X to remove", address)
+		}
+	}
+	return nil
+}
+
+// Removes the validators specified in 'deletes' from validator set 'vals'.
+// Should not fail as verification has been done before.
+func (vals *ValidatorSet) applyRemovals(deletes []*Validator) {
+
+	for _, valUpdate := range deletes {
+		address := valUpdate.Address
+		_, removed := vals.Remove(address)
+		if !removed {
+			// Should never happen
+			panic(fmt.Sprintf("failed to remove validator %X", address))
+		}
+	}
+}
+
+// UpdateWithChangeSet attempts to update the validator set with 'changes'
+// It performs the following steps:
+// - validates the changes making sure there are no duplicates and
+// separates them in updates and deletes
+// - verifies that applying the updates will not result in errors (e.g. overflows)
+// - computes the priorities of validators that are added and/or changed against the final set
+// - verifies that applying the removals will not result in errors
+//   Note: currently an error is issued if a validator to be removed is not present in the set
+// - applies the updates against the validator set and performs scaling of priority values
+// - applies the removals against the validator set and performs scaling and centering of priority values
+// If error is detected during verification steps an error is returned and the validator set
+// is not changed.
+func (vals *ValidatorSet) UpdateWithChangeSet(changes []*Validator) error {
```

Comment:
Beautiful code refactoring. I just have question regarding contract we have with ABCI app, i.e., what happens if an application returns changes which contains duplicates for example?

Reply:
Yes, the error bubbles up:
UpdateWithChangeSet() -> execution.go:updateState() -> execution.go:ApplyBlock() -> reactor.go:poolRoutine() where we panic

We could probably discard the updates, notify the application and continue without the validator set changes.
It is also possible to discard duplicates and invalid changes and apply the rest good ones...do we want to do this?

Reply:
Since we don't have a good way to "notify the application" right now, and since something like duplicate validators probably indicates a serious error in application logic, we should probably treat it as is and have tendermint halt. Of course this means we need to better support rollbacks for devs with buggy apps ... #3196

Reply:
Also to @milosevic earlier point, should I update the ABCI documentation? Part of this PR? Or someone else needs/ wants to do it?

Reply:
Can be done separately, good point though.